vector/src/main/java/im/vector/Matrix.java

```diff
@@ -479,85 +570,138 @@ public static boolean hasValidSessions() {
 
         if (clearCredentials) {
-            mLoginStorage.removeCredentials(session.getHomeServerConfig());
+            mLoginStorage.removeCredentials(tchapSession.getConfig());
         }
 
-        session.getDataHandler().removeListener(mLiveEventListener);
-
         ApiCallback<Void> callback = new SimpleApiCallback<Void>() {
-            @Override
-            public void onSuccess(Void info) {
-                VectorApp.removeSyncingSession(session);
-
-                synchronized (LOG_TAG) {
-                    mMXSessions.remove(session);
+            private void onCompletion() {
+                for (MXSession session : tchapSession.getSessions()) {
+                    session.getDataHandler().removeListener(mLiveEventListener);
+                    VectorApp.removeSyncingSession(session);
                 }
 
+                String sessionKey = tchapSession.getMainSession().getMyUserId();
+                mTchapSessionKeys.remove(sessionKey);
+                mTchapSessions.remove(sessionKey);
+
                 if (null != aCallback) {
                     aCallback.onSuccess(null);
                 }
             }
+
+            @Override
+            public void onSuccess(Void info) {
+                MXSession shadowSession = tchapSession.getShadowSession();
+                if (shadowSession != null) {
+                    if (clearCredentials) {
+                        shadowSession.logout(context, new SimpleApiCallback<Void>(aCallback) {
+                            @Override
+                            public void onSuccess(Void info) {
+                                onCompletion();
+                            }
+                        });
+                    } else {
+                        shadowSession.clear(context, new SimpleApiCallback<Void>(aCallback) {
+                            @Override
+                            public void onSuccess(Void info) {
+                                onCompletion();
+                            }
+                        });
+                    }
+                } else {
+                    onCompletion();
+                }
+
+            }
         };
 
         if (clearCredentials) {
-            session.logout(context, callback);
+            tchapSession.getMainSession().logout(context, callback);
         } else {
-            session.clear(context, callback);
+            tchapSession.getMainSession().clear(context, callback);
         }
 
         // Clear media scan database
         // TODO The media scan database clear should be handled during the media cache clear when the MediaScanManager will be moved into the sdk.
         Realm realm = Realm.getDefaultInstance();
-        MediaScanManager mediaScanManager = new MediaScanManager(session.getMediaScanRestClient(), realm);
+        MediaScanManager mediaScanManager = new MediaScanManager(tchapSession.getMainSession().getMediaScanRestClient(), realm);
```

Comment:
shadow session does not have MediaScan manager?

Reply:
shadow session has a MediaScan manager but it is based on the same Realm DB
so only one 'clearAntiVirusScanResults' is required here

I agree this looks strange, and this should be improved later

Reply:
ok

Reply:
FYI I updated the code by using directly a MediaScanDao instance